src/ui_fsmenu/main.cc

```diff
@@ -19,184 +19,564 @@
+	      .str());
+}
+
+void FullscreenMenuMain::set_button_visibility(const bool v) {
+	if (visible_ == v) {
+		return;
+	}
+	visible_ = v;
+	vbox1_.set_visible(v);
+	vbox2_.set_visible(v);
+	copyright_.set_visible(v);
+	version_.set_visible(v);
+}
+
+bool FullscreenMenuMain::handle_mousepress(uint8_t, int32_t, int32_t) {
+	if (init_time_ != kNoSplash) {
+		init_time_ = kNoSplash;
+		return true;
+	}
+	return false;
+}
+
+bool FullscreenMenuMain::handle_key(const bool down, const SDL_Keysym code) {
+	if (down) {
+		bool fell_through = false;
+		if (init_time_ != kNoSplash) {
+			init_time_ = kNoSplash;
+			fell_through = true;
+		}
+		switch (code.sym) {
+		case SDLK_ESCAPE:
+			if (!fell_through) {
+				end_modal<FullscreenMenuBase::MenuTarget>(FullscreenMenuBase::MenuTarget::kBack);
+				return true;
+			}
+			break;
+		case SDLK_t:
+			end_modal<FullscreenMenuBase::MenuTarget>(FullscreenMenuBase::MenuTarget::kTutorial);
+			return true;
+		case SDLK_c:
+			if (!filename_for_continue_.empty()) {
+				end_modal<FullscreenMenuBase::MenuTarget>(
+				   FullscreenMenuBase::MenuTarget::kContinueLastsave);
+				return true;
+			}
+			break;
+		case SDLK_n:
+			end_modal<FullscreenMenuBase::MenuTarget>(FullscreenMenuBase::MenuTarget::kNewGame);
+			return true;
+		case SDLK_h:
+			end_modal<FullscreenMenuBase::MenuTarget>(FullscreenMenuBase::MenuTarget::kCampaign);
+			return true;
+		case SDLK_l:
+			end_modal<FullscreenMenuBase::MenuTarget>(FullscreenMenuBase::MenuTarget::kLoadGame);
+			return true;
+		case SDLK_j:
+			end_modal<FullscreenMenuBase::MenuTarget>(FullscreenMenuBase::MenuTarget::kMetaserver);
+			return true;
+		case SDLK_u:
+			end_modal<FullscreenMenuBase::MenuTarget>(
+			   FullscreenMenuBase::MenuTarget::kOnlineGameSettings);
+			return true;
+		case SDLK_p:
+			end_modal<FullscreenMenuBase::MenuTarget>(FullscreenMenuBase::MenuTarget::kLan);
+			return true;
+		case SDLK_e:
+			end_modal<FullscreenMenuBase::MenuTarget>(FullscreenMenuBase::MenuTarget::kEditor);
+			return true;
+		/* case SDLK_a:
+			end_modal<FullscreenMenuBase::MenuTarget>(FullscreenMenuBase::MenuTarget::kAddOns);
+			return true; */
+		case SDLK_o:
+			end_modal<FullscreenMenuBase::MenuTarget>(FullscreenMenuBase::MenuTarget::kOptions);
+			return true;
+		case SDLK_r:
+			end_modal<FullscreenMenuBase::MenuTarget>(FullscreenMenuBase::MenuTarget::kReplay);
+			return true;
+		case SDLK_F1:
+			end_modal<FullscreenMenuBase::MenuTarget>(FullscreenMenuBase::MenuTarget::kAbout);
+			return true;
+		case SDLK_s:
+			singleplayer_.toggle();
+			return true;
+		case SDLK_m:
+			multiplayer_.toggle();
+			return true;
+		default:
+			break;
+		}
+	}
+	return UI::Panel::handle_key(down, code);
+}
+
+// Position the image at the screen center, as large as possible without upscaling
+Rectf FullscreenMenuMain::image_pos(const Image& i) {
+	const float w = i.width();
+	const float h = i.height();
+	if (w < get_w() || h < get_h()) {
```

Comment:
Same scaling algorithm as in the progress window. Can they be pulled out sensibly into a common function?

"no" is a perfectly good answer.